JournalInconsistentException: failed to load journal type 100

<!-- (At least include the following, feel free to add if you have more content) -->

### Real behavior (Required)
```
2023-03-22 17:27:45,815 WARN (replayer|90) [GlobalStateMgr$5.runOneCycle():1848] got interrupt exception or inconsistent exception when replay journal 6974, will exit,
com.starrocks.journal.JournalInconsistentException: failed to load journal type 100
        at com.starrocks.persist.EditLog.loadJournal(EditLog.java:1031) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1987) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$5.runOneCycle(GlobalStateMgr.java:1845) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$5.run(GlobalStateMgr.java:1910) [starrocks-fe.jar:?]
Caused by: java.lang.IllegalStateException: 
        at com.google.common.base.Preconditions.checkState(Preconditions.java:510) ~[spark-dpp-1.0.0.jar:?]
        at com.starrocks.transaction.LakeTableTxnLogApplier.applyVisibleLog(LakeTableTxnLogApplier.java:63) ~[starrocks-fe.jar:?]
        at com.starrocks.transaction.DatabaseTransactionMgr.updateCatalogAfterVisible(DatabaseTransactionMgr.java:1524) ~[starrocks-fe.jar:?]
        at com.starrocks.transaction.DatabaseTransactionMgr.replayUpsertTransactionState(DatabaseTransactionMgr.java:1613) ~[starrocks-fe.jar:?]
        at com.starrocks.transaction.GlobalTransactionMgr.replayUpsertTransactionState(GlobalTransactionMgr.java:627) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.loadJournal(EditLog.java:602) ~[starrocks-fe.jar:?]
        ... 4 more
```

### StarRocks version (Required)
 branch-3.0 3209c7c

--- a/fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java
+++ b/fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java
@@ -49,6 +49,7 @@ import com.starrocks.lake.LakeTable;
 import com.starrocks.lake.LakeTablet;
 import com.starrocks.lake.ShardDeleter;
 import com.starrocks.lake.Utils;
+import com.starrocks.persist.EditLog;
 import com.starrocks.persist.gson.GsonUtils;
 import com.starrocks.server.GlobalStateMgr;
 import com.starrocks.sql.optimizer.statistics.IDictManager;
@@ -76,6 +77,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
@@ -241,6 +243,11 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
     @VisibleForTesting
     public static void writeEditLog(LakeTableSchemaChangeJob job) {
         GlobalStateMgr.getCurrentState().getEditLog().logAlterJob(job);
+    }
+
+    @VisibleForTesting
+    public static Future<Boolean> writeEditLogAsync(LakeTableSchemaChangeJob job) {
+        return GlobalStateMgr.getCurrentState().getEditLog().logAlterJobNoWait(job);
     }
 
     @VisibleForTesting
@@ -451,6 +458,8 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
             return;
         }
 
+        long startWriteTs;
+        Future<Boolean> editLogFuture;
         // Replace the current index with shadow index.
         Set<String> modifiedColumns;
         List<MaterializedIndex> droppedIndexes;
@@ -465,7 +474,25 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
             modifiedColumns = collectModifiedColumnsForRelatedMVs(table);
             // Below this point, all query and load jobs will use the new schema.
             droppedIndexes = visualiseShadowIndex(table);
+
+            // update colocation info and inactivate related mv
+            try {
+                GlobalStateMgr.getCurrentColocateIndex().updateLakeTableColocationInfo(table, true, null);
+            } catch (DdlException e) {
+                // log an error if update colocation info failed, schema change already succeeded
+                LOG.error("table {} update colocation info failed after schema change, {}.", tableId, e.getMessage());
+            }
+
+            inactiveRelatedMv(modifiedColumns, table);
+
+            this.jobState = JobState.FINISHED;
+            this.finishedTimeMs = System.currentTimeMillis();
+
+            startWriteTs = System.nanoTime();
+            editLogFuture = writeEditLogAsync(this);
         }
+
+        EditLog.waitInfinity(startWriteTs, editLogFuture);
 
         // Delete tablet and shards
         List<Long> unusedShards = new ArrayList<>();
@@ -475,29 +502,6 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
         // TODO: what if unusedShards deletion is partially successful?
         ShardDeleter.dropTabletAndDeleteShard(unusedShards, GlobalStateMgr.getCurrentStarOSAgent());
-
-        // update colocation info and inactivate related mv
-        try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
-            LakeTable table = (db != null) ? db.getTable(tableId) : null;
-            if (table != null) {
-                try {
-                    GlobalStateMgr.getCurrentColocateIndex().updateLakeTableColocationInfo((OlapTable) table,
-                            true /* isJoin */, null /* expectGroupId */);
-                } catch (DdlException e) {
-                    // log an error if update colocation info failed, schema change already succeeded
-                    LOG.error("table {} update colocation info failed after schema change, {}.", tableId, e.getMessage());
-                }
-
-                inactiveRelatedMv(modifiedColumns, table);
-            } else {
-                LOG.info("database or table has been dropped while trying to update colocation info for job {}.", jobId);
-            }
-        }
-
-        this.jobState = JobState.FINISHED;
-        this.finishedTimeMs = System.currentTimeMillis();
-
-        writeEditLog(this);
 
         if (span != null) {
             span.end();

--- a/fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java
+++ b/fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java
@@ -1088,7 +1088,7 @@ public class EditLog {
     /**
      * wait for JournalWriter commit all logs
      */
-    public void waitInfinity(long startTime, Future<Boolean> task) {
+    public static void waitInfinity(long startTime, Future<Boolean> task) {
         boolean result;
         int cnt = 0;
         while (true) {

--- a/fe/fe-core/src/test/java/com/starrocks/alter/LakeTableSchemaChangeJobTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/alter/LakeTableSchemaChangeJobTest.java
@@ -58,6 +58,7 @@ import com.starrocks.thrift.TStorageMedium;
 import com.starrocks.thrift.TStorageType;
 import mockit.Mock;
 import mockit.MockUp;
+import org.apache.commons.lang3.concurrent.ConcurrentUtils;
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
@@ -68,6 +69,7 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.Future;
 import java.util.stream.Collectors;
 import javax.validation.constraints.NotNull;
 
@@ -732,6 +734,11 @@ public class LakeTableSchemaChangeJobTest {
             @Mock
             public void writeEditLog(LakeTableSchemaChangeJob job) {
                 // nothing to do.
+            }
+
+            @Mock
+            public Future<Boolean> writeEditLogAsync(LakeTableSchemaChangeJob job) {
+                return ConcurrentUtils.constantFuture(true);
             }
 
             @Mock

